partitioner/src/main/java/io/confluent/connect/storage/partitioner/TimeBasedPartitioner.java

```diff
@@ -323,5 +323,15 @@ public Long extract(ConnectRecord<?> record) {
         throw new PartitionException("Error encoding partition.");
       }
     }
+
+    private Long extractTimestampFromString(Object timestampValue) {
+      String timestampValueStr = (String) timestampValue;
+      try {
+        return Long.valueOf(timestampValueStr);
+      } catch (NumberFormatException e) {
```

Comment:
My main (big) concern is now we pay a tax when converting timestamps that are not of type long. 
The branching based on exception handling will be way more expensive. 

The best I can think of and confirmed with a research online is to use a regex. We probably don't want to introduce a dep to commons-lang (or similar) just for that.

Reply:
We can just easily avoid this issue by catching the exception from `dateTime.parseMillis(timestampValueStr)` as that is the more common use case and that way we don't "pay the tax" (most of the time).

Reply:
I did some performance testing, to understand the performance tax, that we are introducing.

Base case(handling ISO ts only):
|Number of iterations|ISO ts extraction runtime, ms|
|---------------------|-----------------------------|
| 1000000                   | 1385                                        | 
| 10000000                 | 12172                                      | 
| 100000000               | 121982                                   | 

No regex:
|Number of iterations|ISO ts extraction runtime, ms|Numeric ts extraction runtime, ms|
|---------------------|-----------------------------|---------------------------------|
| 1000000                   | 4253                                        | 45                                                  |
| 10000000                 | 41792                                      | 305                                                |
| 100000000               | 437195                                   |  2855                                             |

Using a regex:
|Number of iterations|ISO ts extraction runtime, ms|Numeric ts extraction runtime, ms|
|---------------------|-----------------------------|---------------------------------|
| 1000000                   | 1962                                        | 273                                                |
| 10000000                 | 13295                                     | 1885                                               |
| 100000000               | 123757                                   | 15437                                             |

Looks like using a regex makes the numeric timestamp extraction slower, but keeps existing ISO timestamp extraction within a small delta.